misc/benchmark/bench.py

```diff
@@ -0,0 +1,259 @@
+#!/usr/bin/env python3
+"""
+    bench.py references the repo[https://github.com/nydusaccelerator/hello-bench].
+"""
+import copy
+import json
+import logging
+import os
+import posixpath
+import subprocess
+import sys
+import time
+import urllib.request
+from contextlib import contextmanager
+from datetime import datetime
+from io import TextIOWrapper
+
+
+def logging_setup(logging_stream=sys.stderr):
+    root = logging.getLogger()
+
+    if root.hasHandlers():
+        return
+
+    verbose = True
+
+    handler = logging.StreamHandler(logging_stream)
+
+    if verbose:
+        root.setLevel(logging.DEBUG)
+        handler.setLevel(logging.DEBUG)
+    else:
+        root.setLevel(logging.INFO)
+        handler.setLevel(logging.INFO)
+
+    formatter = logging.Formatter(
+        "[%(asctime)s] %(levelname)s "
+        "[%(module)s - %(lineno)s:%(funcName)s] "
+        "- %(message)s"
+    )
+    handler.setFormatter(formatter)
+    root.addHandler(handler)
+
+
+logging_setup()
+
+
+def run(cmd, wait: bool = True, verbose=True, **kwargs):
+
+    shell = kwargs.pop("shell", False)
+    if shell:
+        cmd = " ".join(cmd)
+
+    if verbose:
+        logging.info(cmd)
+    else:
+        logging.debug(cmd)
+
+    popen_obj = subprocess.Popen(cmd, shell=shell, **kwargs)
+    if wait:
+        popen_obj.wait()
+    return popen_obj.returncode, popen_obj
+
+
+def get_current_time():
+    return datetime.now()
+
+
+def delta_time(t_end, t_start):
+    delta = t_end - t_start
+    return delta.total_seconds(), delta.microseconds
+
+
+@contextmanager
+def timer(cmd):
+    start = get_current_time()
+    try:
+        rc = os.system(cmd)
+        assert rc == 0
+        end = get_current_time()
+        sec, usec = delta_time(end, start)
+        yield sec + usec / 1e6
+        logging.info("%s, Takes time %u.%u seconds", cmd, sec, usec)
+    finally:
+        pass
+
+
+class RunArgs:
+    def __init__(
+        self, waitURL=""
+    ):
+        self.waitURL = waitURL
+
+
+class Bench:
+    def __init__(self, name, category="other"):
+        self.name = name
+        self.category = category
+
+    def __str__(self):
+        return json.dumps(self.__dict__)
+
+    def set_tag(self, tag):
+        self.name = f"{self.name}:{tag}"
+
+
+class BenchRunner:
+    CMD_URL_WAIT = {
+        "wordpress": RunArgs(waitURL="http://localhost:80"),
```

Comment:
Is it possible to load this info from a file so we can add more similar cases easily?

Reply:
Yes, we can do that. Actually in this version, we only benchmark for wordpress. You can see https://github.com/dragonflyoss/image-service/issues/1220, we will add more images and configs to enhancement this feature.

Reply:
Got it.